카드 수정 API 구현

Park-SooYeon opened the pull request:
# 카드 수정 API 구현

## 📎 해당 이슈

이슈 링크 (#199)

## 🛠 구현 내용 

- 카드 수정 API 구현
- 요청 성공 시, 200 응답
- 현재 보드에 속한 리스트가 아닌 리스트로 이동힐 경우, 409 응답
- 카드 수정 권한이 없을 경우, 403 응답
- 수정할 카드를 찾을 수 없는 경우, 404 응답

## 🙋‍ 리뷰어 참고 사항 

현재 보드에 속한 리스트가 아닌 리스트로 이동힐 경우 어떤 응답이 좋을지 고민하다가..🤔
request body의 list id의 값이 잘못된 경우고 요청 수정 없이 동일한 요청을 보내지 말라는 의미를 가진 것이 400 응답이라 생각하여 400 응답으로 했었습니다.
그러나 좀 더 고민해보니 카드가 속하는 리스트가 보드에 속하지 않는 것은 서버가 가진 데이터의 상태와 충돌하는 것으로 볼 수도 있을 것 같아 409 응답이 적절할 것이라고 생각하여 409 응답으로 수정했습니다.
다른 분들은 어떤 응답이 적절할 것 같다고 생각하시나요? 다른 분들 의견이 궁금합니다.👀

Park-SooYeon committed:
[FEAT] 카드 수정하는 서비스 함수 추가

- 카드 정보를 수정하는 modifyCardById 함수 추가

Park-SooYeon committed:
[TEST] modifyCardById 함수 테스트 코드 추가

Park-SooYeon committed:
[FEAT] CardDTO 수정

- validation를 위한 CardDTO 수정
- ValidateIf 데코레이션을 사용하여 필요한 경우에만 validation을 하도록 수정
- validator함수의 groups 매개변수 기본 값을 빈 배열로 설정

Park-SooYeon committed:
[FEAT] PATCH /api/card/{cardId} 라우터 추가

- 서비스 호출 전, CardDTO로 validation 수행
- 요청 성공 시, status 204 응답 반환

Park-SooYeon committed:
[TEST] PATCH /api/card/{cardId} API 테스트 코드 추가

- 요청 성공 시, 204 응답 반환
- 존재하지 않는 카드 수정 시, 404 응답 반환

Park-SooYeon committed:
[FEAT] 카드 수정 API 권한 체크 로직 추가

- 보드에 속한 사용자만 카드를 수정할 수 있도록 권한 체크 로직 추가
- 보드에 속한 사용자가 카드를 수정할 경우, 403 응답 반환

Park-SooYeon committed:
[FEAT] 카드 수정 API 리스트 이동 체크 로직 추가

- 보드에 속한 리스트로만 이동할 수 있도록 리스트 이동 체크 로직 추가
- 이동하려는 리스트가 해당 보드에 속한 리스트가 아닐 경우, 400 응답 반환

Park-SooYeon committed:
[TEST] 카드 수정 API 변경에 따른 테스트 코드 수정 및 추가

- 카드 수정 API 서비스 함수 변경에 따른 테스트 코드 수정
- 카드 수정 API 권한 및 리스트 이동 체크 로직 테스트 코드 추가

Park-SooYeon committed:
[CHORE] 카드 수정 API 응답 상태 코드 변경

- 보드에 속한 리스트로 이동하는 것이 아닐 경우, 응답코드를 400에서 409로 변경
- 응답 상태 코드 변경에 따른 에러 클래스 생성 및 테스트 코드 수정

Park-SooYeon committed:
[FIX] 테스트 오류 수정

- dueDate의 timezone 설정 때문에 테스트가 실패하는 오류 수정

Park-SooYeon committed:
Merge branch 'develop/server' of https://github.com/boostcamp-2020/Project04-B-Whale into feature/patch-card-api

youngxpepp commented on server/src/dao/CustomBoardRepository.js:
```diff
@@ -28,4 +28,15 @@ export class CustomBoardRepository extends BaseRepository {
         return boardIds;
     }
+
+    async checkAccess({ boardId, userId }) {
```
데이터베이스에 직접 접근하는 계층이다 보니까, find exist count 등의 네이밍을 많이 사용하더라구요.
좀 더 적합한 메소드 이름이 없을까요?

Park-SooYeon replied:
보드에 참여중인 유저인지 체크하는거니까.. isExistUserInBoard와 같은 함수명을 사용하는 것이 좀 더 직관적일까요?🤔

youngxpepp replied:
https://jojoldu.tistory.com/516
이런건 참조해보니까 JPA에선 repository를 통해 뭔가 있는지 없는지 확인하려면 exists라는 키워드를 사용하네요!

Park-SooYeon replied:
commit ab84c23
아래 쿼리 빌더 변경이랑 함께 수정했습니다!

youngxpepp commented on server/src/dto/CardDto.js:
```diff
@@ -1,18 +1,36 @@
-    @IsNumber()
     id;
 
+    title;
+
+    content;
+
+    position;
+
+    dueDate;
+
+    listId;
```
밑에 있는거랑 어떤 차이가 있나요? 중복해서 필드를 선언한 이유가 있을까요?

Park-SooYeon replied:
class-validator에서 ValidateIf 예제는 이렇게 사용하고 있더라구요..
그리고 위가 엔티티의 기본이 되는 값이고 아래 쪽에서 엔티티의 기본이 되는 값에 따라 어떤 validation 옵션을 어디에 적용시킬 건지 체크하는거라고 이해해서 이렇게 작성했습니다.
근데 다시보니 결국엔 IsOptional과 차이가 없는 것 같네요..ValidateIf에 대해 좀 더 찾아봐야겠습니다.🤔

youngxpepp replied:
혹시 참조한 문서 알려주실 수 있을까요? 필드의 중복선언이 신기하네요

Park-SooYeon replied:
https://github.com/typestack/class-validator
여기서 참조했습니다! 예제에서는 다른 필드명을 사용하고있어요.

youngxpepp commented on server/src/service/CardService.js:
```diff
@@ -101,4 +104,30 @@ export class CardService extends BaseService {
+
+        if (!card) throw new EntityNotFoundError('card is not exist');
+
+        const { list } = await this.cardRepository.findOne(cardId, { relations: ['list'] });
+        const { board } = await this.listRepository.findOne(list.id, { relations: ['board'] });
```
하나의 쿼리로 합쳐도 될 것 같다는 생각이 드는데, 쿼리 빌더 사용하시는건 어때욥?

Park-SooYeon replied:
쿼리빌더를 사용해도 어짜피 조인이 두번 일어나니 차이가 없을 것 같다고 생각했어요.
조인이 한번씩 수행되는 쿼리를 두번 호출하기보다는 조인이 두번 수행되는 한방 쿼리로 하는 것이 좀 더 효율적이였을까요?🤔

youngxpepp replied:
네트워크 비용이 더 클거 같다는 생각이 들어요.
114번 째 라인이 완료가 되어야 115번이 실행되는 구조라서, 한방 쿼리가 효율적이라고 생각했습니다ㅎㅎ
두 번 조인하는게 같다면, 네트워크 비용을 줄이는게 좋을거 같아요!

youngxpepp reviewed (changes requested):
고생하셨습니다! validator 잘 만들어 주셔서 감사합니다ㅎㅎ👍
궁금한 점이 있어서 질문 남겼어요!

dh00n commented on server/src/dao/CustomBoardRepository.js:
```diff
@@ -28,4 +28,15 @@ export class CustomBoardRepository extends BaseRepository {
+            .orWhere('invitation.user_id=:userId', { userId })
+            .getMany();
+
+        return isAuth.length !== 0;
```
똑같이 동작하는 메서드 만들어 놨었는데 또 만드셧네요 ,,!😥

youngxpepp replied:
동훈님 메서드는 service 계층에 있지 않아요? 아마 밑으로 내리신거 같은데 아닌감

Park-SooYeon replied:
아 저는 권한 체크할 때, boolean 값을 반환해주려고 새로 만들었어요!

youngxpepp commented:
409에 대한 의견을 안 드렸었네요. 409 코드는 수연님 덕분에 처음 알았어요. 좋다고 생각합니다ㅎㅎ

dh00n commented:
만약 저희 서비스가 현재 보드에 속한 리스트가 아닌 타 보드의 리스트로 이동하는 기능 지원을 안한다면 
저도 409가 맞다고 생각합니다 ㅎㅎ

Park-SooYeon committed:
[STYLE] 리뷰에 따라 코드 수정

- checkAccess 함수명을 좀 더 직관적인 existUserByBoardId로 변경
- cardId로 board를 조회하는 쿼리 추가
- isListOfBoard 함수명을 existListByBoardId로 변경 및 board dao로 이동

youngxpepp reviewed (dismissed):
고생하셨습니다!

dh00n reviewed (dismissed):
👍👍👍👍👍

Park-SooYeon committed:
[FIX] 권한이 없는 사용자가 카드 수정할 수 있는 오류 수정

- 다른 보드에 초대된 유저가 초대되지 않은 다른 보드의 카드를 수정할 수 있는 오류 수정

Park-SooYeon committed:
[TEST] 오류 수정에 따른 테스트 코드 추가

- 초대된 보드가 아닌 보드의 카드를 수정하려한 경우 403을 반환하는지 확인하는 테스트 코드 추가

dh00n reviewed (approved):
👏👏👏👏👏

youngxpepp reviewed (approved):
오류 수정까지 고생하셨습니다!
